src/bundles/communication/MqttController.ts

```diff
@@ -0,0 +1,103 @@
+import { connect, type MqttClient } from 'mqtt/dist/mqtt';
+
+export const STATE_CONNECTED = 'Connected';
+export const STATE_DISCONNECTED = 'Disconnected';
+export const STATE_RECONNECTED = 'Reconnected';
+export const STATE_OFFLINE = 'Offline';
+
+/**
+ * Abstraction of MQTT.
+ *
+ * @param connectionCallback Callback when the connection state changed.
+ * @param messageCallback Callback when a message has been received.
+ */
+export class MqttController {
+  private client: MqttClient | null = null;
+  private connected: boolean = false;
+  private connectionCallback: (status: string) => void;
+  private messageCallback: (topic: string, message: string) => void;
+
+  address: string = '';
+  port: number = 8080;
+
+  constructor(
+    connectionCallback: (status: string) => void,
+    messageCallback: (topic: string, message: string) => void,
+  ) {
+    this.connectionCallback = connectionCallback;
+    this.messageCallback = messageCallback;
+  }
+
+  /**
+   * Sets up MQTT client link and connects to it.
+   * Also handles connection status callbacks.
+   */
+  public connectClient() {
+    if (this.connected || this.address.length === 0) return;
+    let link = `wss://${this.address}:${this.port}/mqtt`;
+    this.client = connect(link);
+    this.connected = true;
+    this.client.on('connect', () => {
+      this.connectionCallback(STATE_CONNECTED);
+    });
+    this.client.on('disconnect', () => {
+      this.connectionCallback(STATE_DISCONNECTED);
+    });
+    this.client.on('reconnect', () => {
+      this.connectionCallback(STATE_RECONNECTED);
+    });
+    this.client.on('offline', () => {
+      this.connectionCallback(STATE_OFFLINE);
+    });
+    this.client.on('message', (topic, message) => {
+      let decoder = new TextDecoder('utf-8');
+      this.messageCallback(topic, decoder.decode(message));
+    });
+  }
+
+  /**
+   * Disconnects the MQTT client.
+   */
+  public disconnect() {
+    if (this.client) {
+      this.client.end(true);
+    }
+    this.connectionCallback = () => {};
+    this.messageCallback = () => {};
+  }
+
+  /**
+   * Broadcasts message to topic.
+   * QoS of 1, all listening devices receive the message at least once.
```

Comment:
Doesn't QoS 1 only guarantee that the sender sends at least once (and the broker receives at least once)? It's up to the QoS between the broker and the subscriber at the point of subscription to determine whether the QoS stays at 1 or gets downgraded between the broker and subscriber.